File: python/232-Implement-Queue-using-Stacks.py
#region EfficientPushAmortizedPop
class MyQueue:

    def __init__(self):
        self.main = list()
        self.holding = list()
        self.front = int

    # Time O(1)
    # Space O(n) Need space for each elem
    def push(self, x: int) -> None:
        if not self.main:
            self.front = x
        self.main.append(x)


    # Time O(1) Amortized
    # Space O(1)
    def pop(self) -> int:
        # Moving elements to the holding stack reverses the order s.t. the top holding stack elem
        # will be the first "queue" elem. Only should move elems to holding stack once empty
        # to maintain order
        if not self.holding:
            while self.main:
                self.holding.append(self.main.pop())
        return self.holding.pop()

    # Time O(1)
    # Space O(1)
    def peek(self) -> int:
        if self.holding:
            return self.holding[len(self.holding) - 1]
        return self.front

    # Time O(1)
    # Space O(1)
    def empty(self) -> bool:
        return not self.main and not self.holding


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

#endregion


#region LessEfficientPush
class MyQueue:

    def __init__(self):
        self.main = list()
        self.holding = list()

    # Time O(n) Each elem is pushed and popped twice except for new one
    # Space O(n) Need space for each elem
    def push(self, x: int) -> None:
        while self.main:
            self.holding.append(self.main.pop())
        self.main.append(x)
        while self.holding:
            self.main.append(self.holding.pop())

    # Time O(1)
    # Space O(1)
    def pop(self) -> int:
        return self.main.pop()

    # Time O(1)
    # Space O(1)
    def peek(self) -> int:
        if self.main:
            return self.main[len(self.main) - 1]

    # Time O(1)
    # Space O(1)
    def empty(self) -> bool:
        return len(self.main) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

#endregion

File: c#/232-Implement-Queue-using-Stacks.cs
// Push: O(1)
// Pop: O(1) amortized -- generally will be O(1) unless qStack empty, then O(n)
// Peek: O(1) same as pop
// Empty: O(1)
public class MyQueue
{
    // Push: we can always push to same stack
    // Pop: we will rely on second stack to simulate a queue. if it's empty, pop all elems from push stack to queue stack. pop top of stack for first elem
    // Peek: peek is same as push except we peek first elem
    // Empty: true if both stack counts are 0

    Stack<int> pushStack;
    Stack<int> qStack;

    public MyQueue()
    {
        pushStack = new();
        qStack = new();
    }

    public void Push(int x)
    {
        pushStack.Push(x);
    }

    public int Pop()
    {
        PrepQueue();
        return qStack.Pop();
    }

    public int Peek()
    {
        PrepQueue();
        return qStack.Peek();
    }

    private void PrepQueue()
    {
        if (qStack.Count == 0)
        {
            while (pushStack.Count > 0)
                qStack.Push(pushStack.Pop());
        }
    }

    public bool Empty()
    {
        return qStack.Count == 0 && pushStack.Count == 0;
    }
}
